pip/pip-420.md

```diff
@@ -0,0 +1,304 @@
+The encode method returns an `EncodeData` object that contains the encoded byte array and schema ID.
+The customized external schemas can set the `SchemaInfoProvider` and retrieve the configs from it,
+extends the interface `AutoCloseable` to support close external schema resources.
+
+```java
+import java.util.concurrent.CompletableFuture;
+
+public interface Schema extends Cloneable {
+
+  /**
+   * Encodes the message into a byte array using the schema.
+   *
+   * @param topic the topic for which the message is being encoded
+   * @param message the message to encode
+   * @return the encoded byte array and schema ID
+   * @throws SchemaSerializationException if the encoding fails
+   */
+  default EncodeData encode(String topic, T message) {
+    return new EncodeData(encode(message), null);
+  }
+
+  default T decode(String topic, ByteBuffer data, byte[] schemaId) {
+    return decode(topic, getBytes(data), schemaId);
+  }
+
+  /**
+   * Decodes a byte array into a message using the schema.
+   *
+   * @param topic the topic for which the message is being decoded
+   * @param data the byte array to decode
+   * @param schemaId the schema ID associated with the data
+   * @return the decoded message
+   * @throws SchemaSerializationException if the decoding fails
+   */
+  default T decode(String topic, byte[] data, byte[] schemaId) {
+    return decode(data, schemaId);
+  }
+
+  default void closeAsync() {
+    return CompletableFuture.completedFuture(null);
+  }
+
+}
+```
+
+# Pulsar Function
+
+For support using third-party schema registry service in Pulsar Function,
+- Support the `SchemaType.EXTERNAL` schema type in Pulsar Function
+
+# Pulsar-GEO replication impact
+
+Integrating third-party schema registry services introduces a new approach to managing schemas for geo-replicated topics.
+
+In the current Pulsar architecture:
+- Schema definitions are stored and managed by the Pulsar brokers.
+- During geo-replication, schema information must also be replicated across clusters to ensure schema consistency.
+
+By using an external schema registry:
+- **Schema management is fully decoupled from Pulsar brokers and replication mechanisms.**
+- This eliminates the need for synchronizing schema data between Pulsar clusters, simplifying geo-replication processes.
+- Supports a unified schema registry for cross-cluster producers and consumers
+
+# Backward & Forward Compatibility
+
+The new schema type `SchemaType.EXTERNAL` doesn't break any existing Pulsar topics, it's not compatible with other Pulsar schema types.
+
+# Alternatives
+
+Use `bytes` schema for "external" schemas, it can't provide any compatibility checks to protect topic data that use Pulsar's native schema system.
+
+# General Notes
+
+# Links
+
+<!--
+Updated afterwards
+-->
+* Mailing List discussion thread: https://lists.apache.org/thread/olx4xm8cdy43omp5c0jm44sj1gp0grcr
+* Mailing List voting thread: https://lists.apache.org/thread/vhq6ox4nh2rx59yoxowftqzv8f9lnm4q
```

Comment:
This link is wrong, please update the correct link

Reply:
update